File: apps/ai-service/app/schemas/canvas.py
"""
Pydantic mirrors of packages/shared-types/src/canvas.ts.
Keep field names in sync (TypeScript uses camelCase — ConfigDict aliases handle that).
"""

from datetime import datetime
from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Field

CANVAS_SCHEMA_VERSION = 1


class _CamelModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=None)


class Vector2(_CamelModel):
    x: float
    y: float


class _BaseShape(_CamelModel):
    id: str
    position: Vector2
    rotation: float = 0.0
    scale: Vector2 = Field(default_factory=lambda: Vector2(x=1, y=1))
    opacity: float = 1.0
    draggable: bool = True
    z_index: int = Field(0, alias="zIndex")


class RectShape(_BaseShape):
    kind: Literal["rect"] = "rect"
    width: float
    height: float
    fill: str
    stroke: str | None = None
    stroke_width: float | None = Field(None, alias="strokeWidth")
    corner_radius: float | None = Field(None, alias="cornerRadius")


class CircleShape(_BaseShape):
    kind: Literal["circle"] = "circle"
    radius: float
    fill: str
    stroke: str | None = None
    stroke_width: float | None = Field(None, alias="strokeWidth")


class LineShape(_BaseShape):
    kind: Literal["line"] = "line"
    points: list[float]
    stroke: str
    stroke_width: float = Field(..., alias="strokeWidth")
    closed: bool | None = None


class TextShape(_BaseShape):
    kind: Literal["text"] = "text"
    text: str
    font_size: float = Field(..., alias="fontSize")
    font_family: str = Field(..., alias="fontFamily")
    fill: str
    width: float | None = None
    align: Literal["left", "center", "right"] | None = None


class ImageShape(_BaseShape):
    kind: Literal["image"] = "image"
    blob_key: str = Field(..., alias="blobKey")
    width: float
    height: float


class GroupShape(_BaseShape):
    kind: Literal["group"] = "group"
    children: list[str]


Shape = Annotated[
    Union[RectShape, CircleShape, LineShape, TextShape, ImageShape, GroupShape],
    Field(discriminator="kind"),
]


class CanvasViewport(_CamelModel):
    pan: Vector2
    zoom: float = 1.0


class CanvasDocument(_CamelModel):
    schema_version: int = Field(CANVAS_SCHEMA_VERSION, alias="schemaVersion")
    id: str
    owner_id: str = Field(..., alias="ownerId")
    name: str
    width: int
    height: int
    background: str
    viewport: CanvasViewport
    shapes: list[Shape]
    created_at_utc: datetime = Field(..., alias="createdAtUtc")
    updated_at_utc: datetime = Field(..., alias="updatedAtUtc")

    def to_summary(self) -> dict[str, Any]:
        """Compact summary for LLM prompts."""
        return {
            "name": self.name,
            "size": {"w": self.width, "h": self.height},
            "shape_count": len(self.shapes),
            "shape_kinds": [s.kind for s in self.shapes],
        }

File: packages/shared-types/src/canvas.ts
/**
 * Canvas JSON state — source of truth.
 *
 * Consumed by:
 *   - apps/web (React-Konva renderer + Zustand store)
 *   - apps/api (mirrored as C# records in Acme.Application/Canvas/)
 *   - apps/ai-service (mirrored as Pydantic models in app/schemas/canvas.py)
 *
 * Versioning: bump CANVAS_SCHEMA_VERSION on any breaking change and add a
 * migration in apps/web/src/lib/canvas/migrate.ts.
 */

export const CANVAS_SCHEMA_VERSION = 1;

export type ShapeId = string;

export type ShapeKind =
  | "rect"
  | "circle"
  | "line"
  | "text"
  | "image"
  | "group";

export interface Vector2 {
  x: number;
  y: number;
}

export interface BaseShape {
  id: ShapeId;
  kind: ShapeKind;
  position: Vector2;
  rotation: number; // degrees
  scale: Vector2;
  opacity: number; // 0..1
  draggable: boolean;
  zIndex: number;
}

export interface RectShape extends BaseShape {
  kind: "rect";
  width: number;
  height: number;
  fill: string;
  stroke?: string;
  strokeWidth?: number;
  cornerRadius?: number;
}

export interface CircleShape extends BaseShape {
  kind: "circle";
  radius: number;
  fill: string;
  stroke?: string;
  strokeWidth?: number;
}

export interface LineShape extends BaseShape {
  kind: "line";
  points: number[]; // [x1,y1,x2,y2,...]
  stroke: string;
  strokeWidth: number;
  closed?: boolean;
}

export interface TextShape extends BaseShape {
  kind: "text";
  text: string;
  fontSize: number;
  fontFamily: string;
  fill: string;
  width?: number;
  align?: "left" | "center" | "right";
}

export interface ImageShape extends BaseShape {
  kind: "image";
  /** Blob URL — relative to AZURE_BLOB_CONTAINER. */
  blobKey: string;
  width: number;
  height: number;
}

export interface GroupShape extends BaseShape {
  kind: "group";
  children: ShapeId[];
}

export type Shape =
  | RectShape
  | CircleShape
  | LineShape
  | TextShape
  | ImageShape
  | GroupShape;

export interface CanvasViewport {
  pan: Vector2;
  zoom: number;
}

export interface CanvasDocument {
  schemaVersion: number;
  id: string;
  ownerId: string;
  name: string;
  width: number;
  height: number;
  background: string;
  viewport: CanvasViewport;
  shapes: Shape[];
  createdAtUtc: string;
  updatedAtUtc: string;
}

/** Convenience constructor for an empty document. */
export function emptyCanvas(
  id: string,
  ownerId: string,
  name = "Untitled",
): CanvasDocument {
  const now = new Date().toISOString();
  return {
    schemaVersion: CANVAS_SCHEMA_VERSION,
    id,
    ownerId,
    name,
    width: 1920,
    height: 1080,
    background: "#ffffff",
    viewport: { pan: { x: 0, y: 0 }, zoom: 1 },
    shapes: [],
    createdAtUtc: now,
    updatedAtUtc: now,
  };
}
